File: practice_java_program/notpad/SingleTonPattern.java
public class SingleTonPattern{
    public static void main(String[] args) {

	SingleTon obj1 = SingleTon.getInstance();
	SingleTon obj2 = SingleTon.getInstance();

	System.out.println(obj1 == obj2);

	SingletonExample obj3 = SingletonExample.getInstance ();
	SingletonExample obj4 = SingletonExample.getInstance();

	System.out.println(obj3 == obj4);
        
    }
}
class SingleTon{
	private static SingleTon instance = new SingleTon();

	private SingleTon(){
		System.out.println("Singleton constructer called");
	}

	public static SingleTon getInstance(){
		return instance;
	}
}
//thread safe with synchronized singleton mathod
class SingletonExample {

    private static SingletonExample instance;

    private SingletonExample() {}

    public static synchronized SingletonExample getInstance() {
        if (instance == null) {
            instance = new SingletonExample();
        }
        return instance;
    }
}

File: python_oriented_interview/notpad/SingleTonPattern.py
import threading

class SingleTon:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            with cls._lock:
                if not cls._instance:
                    print("Singleton constructor called")
                    cls._instance = super(SingleTon, cls).__new__(cls)
        return cls._instance

    @classmethod
    def getInstance(cls):
        return cls()

# Thread safe singleton matching class SingletonExample
class SingletonExample:
    _instance = None
    _lock = threading.Lock()

    @classmethod
    def getInstance(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = SingletonExample()
        return cls._instance

if __name__ == "__main__":
    obj1 = SingleTon.getInstance()
    obj2 = SingleTon.getInstance()
    print(obj1 is obj2)

    obj3 = SingletonExample.getInstance()
    obj4 = SingletonExample.getInstance()
    print(obj3 is obj4)
